### PR TITLE
ARTEMIS-1300 Deadlock when Core direct deliverying and AMQP receive

### PR DESCRIPTION
**Pull request by clebertsuconic**

### CODE REVIEW

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java [hunk 1]
```diff
@@ -828,6 +834,11 @@ public void removeConsumer(final Consumer consumer) {
             }
          }
 
+         if (consumerList.isEmpty()) {
+            // reset when all consumers are removed
+            supportsDirectDeliver = true;
```
Comment (michaelandrepearce) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueImpl.java]: should maybe iterate the consumers list and set to true if all consumers support direct? e.g. was 3 consumers, 2 core and 1 amqp, then the amqp one removes, leaving 2 core would you want it to go back to direct.

e.g. (pseudo)
```
boolean supportsDirectDeliver = true;
for (consumer : consumerList) {
   if (!consumer.supportsDirectDelivery()){
      supportsDirectDeliver = false;
   } 
}
this.supportsDirectDeliver = supportsDirectDeliver;
```

File: artemis-server/src/main/java/org/apache/activemq/artemis/core/server/Consumer.java [hunk 2]
```diff
@@ -19,9 +19,18 @@
 import java.util.List;
 
 import org.apache.activemq.artemis.core.filter.Filter;
+import org.apache.activemq.artemis.spi.core.protocol.SessionCallback;
 
 public interface Consumer {
 
+   /** i
```
Comment (michaelandrepearce) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/server/Consumer.java]: typo i assume

Comment (clebertsuconic) [inline artemis-server/src/main/java/org/apache/activemq/artemis/core/server/Consumer.java]: thanks!

**Comment by clebertsuconic** [comment]
done ^^ amended... 


This PR is very safe to merge. no test regressions on the whole testsuite.

**Comment by michaelandrepearce** [comment]
Just update, We haven't run full tests, but managed to test this with original scenario test that this was causing lock up on, for this case it seems like it no longer locks up.

**Pull request merged**